Add new sample for graphics pipeline libraries

SaschaWillems opened the pull request:
## Description

This PR adds a sample for the new `VK_EXT_graphics_pipeline_library` extension. The sample creates a pipeline library for shared pipeline state parts and reuses this to generate randomized fragment shaders in a background thread. The sample also includes a small tutorial.

I have tested the sample on Windows 10 with an NVIDIA RTX 2060 and recent Vulkan developer driver.

Fixes #441

## General Checklist:

Please ensure the following points are checked:

- [x] My code follows the [coding style](https://github.com/KhronosGroup/Vulkan-Samples/tree/master/CONTRIBUTING.md#Code-Style)
- [x] I have reviewed file [licenses](https://github.com/KhronosGroup/Vulkan-Samples/tree/master/CONTRIBUTING.md#Copyright-Notice-and-License-Template)
- [x] I have updated the header Copyright to reflect the current year (CI build will fail if Copyright is out of date)
- [x] I have commented any added functions (in line with Doxygen)
- [x] I have commented any code that could be hard to understand
- [x] My changes do not add any new compiler warnings
- [x] My changes do not add any new validation layer errors or warnings
- [x] I have used existing framework/helper functions where possible
- [x] My changes build on Windows, Linux, macOS and Android. Otherwise I have [documented any exceptions](https://github.com/KhronosGroup/Vulkan-Samples/tree/master/CONTRIBUTING.md#General-Requirements)
- [x] My changes do not add any regressions
- [x] I have tested every sample to ensure everything runs correctly
- [x] This PR describes the scope and expected impact of the changes I am making

## Sample Checklist

If your PR contains a new or modified sample, these further checks must be carried out *in addition* to the General Checklist:
- [x] I have tested the sample on at least one compliant Vulkan implementation
- [x] If the sample is vendor-specific, I have [tagged it appropriately](https://github.com/KhronosGroup/Vulkan-Samples/tree/master/CONTRIBUTING.md#General-Requirements)
- [x] I have stated on what implementation the sample has been tested so that others can test on different implementations and platforms
- [x] Any dependent assets have been merged and published in downstream modules
- [x] For new samples, I have added a paragraph with a summary to the appropriate chapter in the [samples readme](./../samples/README.md)

SaschaWillems committed:
Updated Vulkan Headers to 1.3.211

SaschaWillems committed:
First working version of the graphics pipeline library sample

Work-in-progress

SaschaWillems committed:
Descriptor pool

SaschaWillems committed:
Moved pipeline creation to thread

SaschaWillems committed:
Display timing informations

SaschaWillems committed:
Shader with different lighting paths and random colors

Code cleanup

SaschaWillems committed:
Use separate pipeline cache

SaschaWillems committed:
Added graphics pipeline sample to samples listing readme

SaschaWillems committed:
Comments and code cleanup

SaschaWillems committed:
Slightly improved sample. Automatically scale viewports with pipeline count, additional lighting model, fixes, etc.

SaschaWillems committed:
Random colors

SaschaWillems committed:
Added readme for graphics pipeline library sample

SaschaWillems committed:
Adde some performance numbers to the readme

SaschaWillems committed:
Clang format

SaschaWillems committed:
Removed note on pipeline creation time, as it was heavily influenced by runtime GLSL compilation and not representative of the actual extension improvements

marty-johnson59 commented:
Added reviewer suggestions.  Would be great if we could get the DT folks (@aurolou. @mklefrancois) to review as I believe that's the target for this EXT.  Thanks!

gary-sweet reviewed (commented):
Builds ok for me, and correctly reports that the required extensions are not supported, but that's all I can test I'm afraid.

gpx1000 reviewed (dismissed):
works well here Linux

TomAtkinsonArm commented on samples/README.md:
````diff
@@ -227,6 +227,10 @@ sample rate map as a base for the next frame.
 
+### [Graphics pipeline library](./extensions/graphics_pipeline_library)<br/>
+**Extension**: [```VK_EXT_graphics_pipeline_library```](https://www.khronos.org/registry/vulkan/specs/1.3-extensions/man/html/VK_EXT_graphics_pipeline_library.html) <br/>
+Uses the graphics pipeline library extensions to improve run-time pipeline creation. Instead of creating the whole pipeline at once, this sample makes use of that extension to pre-built shared pipeline parts like like vertex input state and fragment output state. These building blocks are then used to create pipelines at runtime, improving build times compared to traditional pipeline creation.
````
`pre-built shared pipeline parts like like vertex input state`

to

`pre-build shared pipeline parts like vertex input state`

SaschaWillems replied:
Thanks for spotting this. I have pushed a change to fix this.

TomAtkinsonArm reviewed (commented):
Small typo. Extension not supported, I will update graphics drivers and try again

SaschaWillems committed:
Fixed spelling

Tobski commented on samples/extensions/graphics_pipeline_library/README.md:
```diff
@@ -0,0 +1,140 @@
+
+This will result in the pipeline state object to be used at draw time. 
+
+A note on `VK_PIPELINE_CREATE_LINK_TIME_OPTIMIZATION_BIT_EXT`: This is an optimization flag. If specified, implementations are allowed to do additional optimization passes. This may increase build times but can in turn result in lower runtime costs.
```
I'd suggest it's worth mentioning here that we suggest apps build first without this flag, and then build again with this flag enabled on the slow path, and switch it out for the fast-linked version when available.

Tobski replied:
I also wonder if it's worth making the sample behave like this? The pieces to do it are all there, so I don't think it'd be a big change.

SaschaWillems replied:
That's indeed a good point. I don't want the sample to become too complicated, so would it be okay if there was a checkbox in the UI where the optimization flag can be toggled? That would let people actually see the difference in build time very easily.

Tobski replied:
That's probably a reasonable compromise.

Tobski commented on samples/extensions/graphics_pipeline_library/graphics_pipeline_library.cpp:
```diff
@@ -0,0 +1,537 @@
+	vkUpdateDescriptorSets(get_device().get_handle(), static_cast<uint32_t>(write_descriptor_sets.size()), write_descriptor_sets.data(), 0, nullptr);
+}
+
+// Compiling shaders can be simplified with the new extension, so we only require code to generate the SPIR-V in thhis sample
```
```suggestion
// Compiling shaders can be simplified with the new extension, so we only require code to generate the SPIR-V in this sample
```

Tobski commented on samples/extensions/graphics_pipeline_library/README.md:
```diff
@@ -0,0 +1,140 @@
+
+## Conclusion
+
+With the new extension it's now possible to separate the monolithic pipeline state into multiple parts that can be r-used and built independently. This opens up new possibilities for optimizing pipeline creation and reducing hick-ups at runtime.
```
```suggestion
With the new extension it's now possible to separate the monolithic pipeline state into multiple parts that can be reused and built independently. This opens up new possibilities for optimizing pipeline creation and reducing hitches at runtime.
```

Hitches is usually the term used by most ISVs I've heard.

Tobski reviewed (changes requested):
Mostly LGTM, some minor comments and one suggested modification.

SaschaWillems committed:
Spelling fixes and wording

SaschaWillems committed:
Fix typo

SaschaWillems committed:
Added UI toggle for link time optimizations

SaschaWillems commented:
@Tobski: Thank you very much for your valuable feedback. I have updated the sample and added an UI toggle for the link time optimization flag and also fixed the readme.

SaschaWillems committed:
Merge branch 'master' into graphics_pipeline_library

# Conflicts:
#	samples/README.md

pdaniell-nv commented on samples/extensions/graphics_pipeline_library/graphics_pipeline_library.cpp:
```diff
@@ -0,0 +1,539 @@
+
+	VkGraphicsPipelineCreateInfo executable_pipeline_create_info{};
+	executable_pipeline_create_info.sType = VK_STRUCTURE_TYPE_GRAPHICS_PIPELINE_CREATE_INFO;
+	executable_pipeline_create_info.pNext = &linking_info;
```
executable_pipeline_create_info.layout must not be left NULL see https://www.khronos.org/registry/vulkan/specs/1.3-extensions/html/vkspec.html#VUID-VkGraphicsPipelineCreateInfo-flags-06729 and https://www.khronos.org/registry/vulkan/specs/1.3-extensions/html/vkspec.html#VUID-VkGraphicsPipelineCreateInfo-flags-06730. This came about due a recent spec change: https://gitlab.khronos.org/vulkan/vulkan/-/merge_requests/5164

SaschaWillems replied:
Thanks for pointing that out. I have updated the sample accordingly.

TomAtkinsonArm commented on samples/README.md:
````diff
@@ -235,6 +235,10 @@ MoltenVk: [MoltenVk](https://github.com/KhronosGroup/MoltenVK).  Also demonstrat
 
+### [Graphics pipeline library](./extensions/graphics_pipeline_library)<br/>
+**Extension**: [```VK_EXT_graphics_pipeline_library```](https://www.khronos.org/registry/vulkan/specs/1.3-extensions/man/html/VK_EXT_graphics_pipeline_library.html) <br/>
+Uses the graphics pipeline library extensions to improve run-time pipeline creation. Instead of creating the whole pipeline at once, this sample makes use of that extension to pre-build shared pipeline parts like like vertex input state and fragment output state. These building blocks are then used to create pipelines at runtime, improving build times compared to traditional pipeline creation.
````
`like like` -> `such as`

SaschaWillems replied:
Thanks for pointing this out. Fixed it.

TomAtkinsonArm reviewed (changes requested):
LGTM - A tiny typo I missed earlier. Cant run as my devices do not support the extension

SaschaWillems committed:
Fixed typp

SaschaWillems committed:
Set pipeline layout

SaschaWillems committed:
Clang format

marty-johnson59 commented:
@SaschaWillems, this has now  had approvals from Tobias, Piers and Steve (tho some are stale).  Looks like all threads are resolved now?  Are we ready to merge?

SaschaWillems commented:
Yes, feel free to merge :)
